modules/generators.py

```diff
@@ -61,6 +68,23 @@ async def _claude_legacy_factory(
                 "content": "No function call was included in the last message. Please include a function call in the next message using the <[tool_name]> [args] </[tool_name]> syntax.",
             }
         )
+
+    # add cache control to last 2 user messages
+    cache_control_breakpoints = 0
+    for msg in reversed(wrapped_messages):
+        if msg["role"] == "user":
+            text_content = msg["content"]
```

Comment:
I might suggest adding an assertion that `text_content` is in fact a string. Otherwise, I expect the agent would get some perhaps-hard-to-diagnose errors from the Anthropic API.